cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -424,14 +444,99 @@ func (c ControllerContext) IsControllerEnabled(controllerDescriptor *ControllerD
-type InitFunc func(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller controller.Interface, enabled bool, err error)
+// NewClientConfig is a shortcut for ClientBuilder.Config. It wraps the error with an additional message.
+func (c ControllerContext) NewClientConfig(name string) (*restclient.Config, error) {
+	config, err := c.ClientBuilder.Config(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client config for %q: %w", name, err)
+	}
+	return config, nil
+}
+
+// NewClient is a shortcut for ClientBuilder.Client. It wraps the error with an additional message.
+func (c ControllerContext) NewClient(name string) (kubernetes.Interface, error) {
+	client, err := c.ClientBuilder.Client(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client for %q: %w", name, err)
+	}
+	return client, nil
+}
+
+// Runnable represents a loop implementing any piece of logic really.
+type Runnable interface {
+	// Run implements any kind of processing loop.
+	// When there is anything to be done, it blocks until the context is cancelled.
+	// Run must ensure all goroutines are terminated before returning.
+	Run(context.Context)
+}
+
+// Controller defines the base interface that all controller wrappers must implement.
+type Controller interface {
+	// Name returns the controller's canonical name.
+	Name() string
+
+	// Runnable is the main controller loop.
+	Runnable
+}
+
+// runnableFunc makes it possible to turn a function into a Runnable quickly.
+type runnableFunc func(context.Context)
+
+func (run runnableFunc) Run(ctx context.Context) {
+	run(ctx)
+}
+
+// runnables is a slice of Runnable objects that are to be run concurrently.
+type runnables []Runnable
+
+func (rx runnables) Run(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(len(rx))
+	for _, r := range rx {
+		go func() {
+			defer wg.Done()
+			r.Run(ctx)
+		}()
+	}
+	wg.Wait()
+}
+
+func runRunnables(ctx context.Context, rx ...Runnable) {
+	runnables(rx).Run(ctx)
+}
+
+// named can be easily embedded to implement Name() for a struct.
+type named string
+
+func (n named) Name() string {
+	return string(n)
+}
+
+func newNamed(name string) named {
+	return named(name)
+}
+
+// namedRunnable implements the Controller interface.
+type namedRunnable struct {
+	named
+	Runnable
+}
+
+func newNamedRunnable(runnable Runnable, name string) *namedRunnable {
+	return &namedRunnable{
+		named:    newNamed(name),
+		Runnable: runnable,
+	}
+}
+
+// newNamedRunnableFunc makes it easy to turn a single function into a Controller.
+func newNamedRunnableFunc(fnc func(context.Context), name string) *namedRunnable {
+	return newNamedRunnable(runnableFunc(fnc), name)
+}
+
+// InitFunc is a constructor for a controller.
+// A nil Controller returned means that the associated controller is disabled.
+type InitFunc func(ctx context.Context, controllerContext ControllerContext, controllerName string) (Controller, error)
```

Comment:
Nit: Doesn't this not really act as an initializer anymore? It returns a controller which then has a runnable function. This is more of a constructor now, maybe this should be renamed? This confused me a bit when looking through since it's called in the builder but doesn't start any routines.

Reply:
Renamed in 439fa60539f28f48a958bb354c83b625b568d8a1